hack/tools/janitor/janitor.go

```diff
@@ -162,12 +204,152 @@ func (s *janitor) deleteVSphereVMs(ctx context.Context, folder string) error {
+			}
+
+			deletionMarkerKey, err = s.vSphereClients.FieldsManager.FindKey(ctx, vSphereDeletionMarkerName)
+			if err != nil && err != object.ErrKeyNameNotFound {
+				return errors.Wrapf(err, "finding custom field %q", vSphereDeletionMarkerName)
+			}
+		}
+	}
+
+	objectsToMark := []*managedElement{}
+	objectsToDelete := []*managedElement{}
+
+	// Filter elements and collect two groups:
+	// * objects to add the timestamp field
+	// * objects to destroy
+	for i := range managedEntities {
+		managedEntity := managedEntities[i]
+
+		// We mark any object we find with a timestamp to determine the first time we did see this item.
+		// This is used as replacement for the non-existing CreationTimestamp on objects.
+		timestamp, err := getDeletionMarkerTimestamp(deletionMarkerKey, managedEntity.entity.Value)
+		if err != nil {
 			return err
 		}
+		// If no timestamp was found: queue it to get marked.
+		if timestamp == nil {
+			objectsToMark = append(objectsToMark, managedEntity)
+			continue
+		}
+
+		// Filter out objects we don't have to cleanup depending on s.maxCreationDate.
+		if timestamp.After(s.maxCreationDate) {
+			log.Info("Skipping deletion of object: marked timestamp does not exceed maxCreationDate", "timestamp", timestamp, "inventoryPath", managedEntity.element.Path)
+			continue
+		}
+
+		// Filter out objects which have children.
+		if hasChildren[managedEntity.element.Path] {
+			log.Info("Skipping deletion of object: object has child objects of a different type", "inventoryPath", managedEntity.element.Path)
+			continue
+		}
+
+		objectsToDelete = append(objectsToDelete, managedEntity)
 	}
+
+	for i := range objectsToMark {
+		managedElement := objectsToMark[i]
+		log.Info("Marking resource object for deletion in vSphere", objectType, managedElement.element.Path)
+
+		if s.dryRun {
+			// Skipping actual mark on dryRun.
+			continue
+		}
+
+		if err := s.vSphereClients.FieldsManager.Set(ctx, managedElement.entity.Reference(), deletionMarkerKey, time.Now().Format(time.RFC3339)); err != nil {
+			return errors.Wrapf(err, "setting field %s on object %s", vSphereDeletionMarkerName, managedElement.element.Path)
+		}
+	}
+
+	// sort objects to delete so children are deleted before parents
+	sort.Slice(objectsToDelete, func(i, j int) bool {
+		a := objectsToDelete[i]
+		b := objectsToDelete[j]
+
+		return strings.Count(a.element.Path, "/") > strings.Count(b.element.Path, "/")
+	})
+
+	destroyTasks := []*object.Task{}
+	for _, managedEntity := range objectsToDelete {
+		log.Info("Destroying object in vSphere", objectType, managedEntity.element.Path)
+		if s.dryRun {
+			// Skipping actual destroy on dryRun.
+			continue
+		}
+
+		task, err := object.NewCommon(s.vSphereClients.Vim, managedEntity.entity.Reference()).Destroy(ctx)
+		if err != nil {
+			return err
+		}
+		log.Info("Created Destroy task for object", objectType, managedEntity.element.Path, "task", task.Reference().Value)
```

Comment:
Isn't this a bit racy? If we delete a child and then a parent element, I assume there is no guarantee in which order vc will delete them?

Reply:
That's right, but also not really an issue.

If you delete a parent resource pool or folder, the childrens get moved one layer up. (also means doing the sort above is not necessarily required)

Reply:
Does this mean one of the delete then fails, but that's not a problem because we delete one run later then?

Reply:
Nope, it will go through/just work and not even fail.